common/notifications.ts

```diff
@@ -1,13 +1,36 @@
 export enum NotificationType {
   EmailVerification = 'emailVerification',
+  PartyInvite = 'partyInvite',
 }
 
-export interface EmailVerificationNotification {
+export interface BaseNotificationData {
+  readonly type: NotificationType
+}
+
+export interface PartyInviteNotificationData extends BaseNotificationData {
+  type: typeof NotificationType.PartyInvite
+  from: string
+  partyId: string
+}
+
+export type NotificationData = PartyInviteNotificationData
+
+export interface BaseNotification {
+  readonly id: string
+  readonly type: NotificationType
+  readonly read: boolean
+  readonly createdAt: number
+  readonly data?: NotificationData
```

Comment:
Even if you remove the type from `NotificationData`, there is still zero guarantee that these things match. In fact, there's even less of a chance that you can guarantee they match at that point in the type system.

Like I said above, I think you're trying way too hard to share these type all the way through. We structure data in the database for reasons of the database: efficiency, query-ability, ability to update/migrate the schema, etc. There's no reason these same constraints apply to all the other ways we use these things, there's no reason that clients should need to deal with the data in a certain way just because it's structured that way in the database. Take the data out of the database, convert it to a form that is actually useful to clients and then send it to them, there's no real reason to share all these structures in this way.

Reply:
Yeah, that's what I did (I haven't pushed the changes yet). I let DB (and server side in general) have its own data structure for the notification, and the one here in the common folder is the one that's sent over the wire by the server and used on the client. I guess I was still sharing the `NotificationData` structure across the stack, which is probably why I had problems with typings when I tried to remove the `type` property in one place, but not the other :d. I'll just have a separate structure for the `NotificationData` that is saved in the DB, and the one defined here in the common folder will be the one with the optional `data` structure and without a `type` property in it.

Is that a good idea, or am I still misunderstanding things? ^^

Reply:
There's two options:
1) Have a container interface for all notifications that can contain some extra data. This container interface is never extended or discriminated on, it's always the same structure. Inside, the `data` field is a union type that can be discriminated
2) Have a base interface that has a `type` field that all notifications extend from. The notification type is then a union of all those notification types, and extra fields for a particular notification type are placed directly on this interface (*not* in a `data` field)

It sounds like you're doing some combination of these two approaches and I highly do not recommend that, if you do that you make it so that we have to cast `NotificationData` every single time because there's nothing to discriminate it with. We want a system here that does not require casting, that is the goal